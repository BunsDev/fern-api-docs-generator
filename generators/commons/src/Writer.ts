import { AstNode } from "./AstNode";

export declare namespace Writer {
    interface Args {
        tabSize?: number;
    }
}

export class Writer {
    private buffer: string;
    private indentLevel: number;
    private hasWrittenAnything: boolean;
    private indentSize: number;

    constructor(indentSize: number) {
        this.buffer = "";
        this.indentLevel = 0;
        this.hasWrittenAnything = false;
        this.indentSize = indentSize;
    }

    public write(text: string): void {
        if (this.hasWrittenAnything) {
            this.buffer += "\n";
        }

        const indent = this.getIndentString(this.indentSize);
        const indentedText = indent + text.replace(/\n/g, `\n${indent}`);

        this.buffer += indentedText;
        this.hasWrittenAnything = true;
    }

    public newLine(): void {
        this.buffer += "\n";
    }

    public openBlock(
        titles: (string | undefined)[],
        openingCharacter: string | undefined = "{",
        callback: () => void,
        closingCharacter: string | undefined = "}"
    ): void {
        const filteredTitles = titles.filter(title => title !== undefined).join(" ");
        if (filteredTitles) {
            this.write(`${filteredTitles} ${openingCharacter ?? ""}`);
        } else {
            this.write(openingCharacter ?? "");
        }
        
        try {
            this.indent(callback);
        } finally {
            this.write(closingCharacter ?? "");
        }
    }

    public indent(callback: () => void): void {
        this.indentLevel++;
        try {
            callback();
        } finally {
            this.indentLevel--;
        }
    }

    public writeNode(node: AstNode): void {
        node.write(this);
    }

    public toString(): string {
        return this.buffer;
    }

    private getIndentString(tabSize: number): string {
        return " ".repeat(this.indentLevel * tabSize);
    }

    public writeSnippetWithVariables(template: string, variables: Record<string, string | undefined>): void {
        const filledTemplate = template.replace(/\$(\d+)/g, (_, index) => {
            const value = variables[index];
            return value !== undefined ? value : '';
        });

        // Handle removing only placeholders and keeping the rest of the template as is
        const cleanedTemplate = filledTemplate.replace(/(\s*\$\d+\s*)/g, '').replace(/\n\s*\n/g, '\n');
        this.write(cleanedTemplate);
    }
    
}
