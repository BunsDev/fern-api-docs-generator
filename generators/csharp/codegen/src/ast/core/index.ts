export { AstNode } from "./AstNode.js";
export { XmlDocWriter } from "./XmlDocWriter.js";
export { Writer } from "./Writer.js";
