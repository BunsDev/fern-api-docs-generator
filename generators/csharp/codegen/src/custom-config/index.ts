export { BaseCsharpCustomConfigSchema } from "./BaseCsharpCustomConfigSchema.js";
export { validateReadOnlyMemoryTypes } from "./validateReadOnlyMemoryTypes.js";
