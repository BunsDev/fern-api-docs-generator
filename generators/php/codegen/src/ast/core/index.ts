export { AstNode } from "./AstNode.js";
export { GLOBAL_NAMESPACE } from "./Constant.js";
export { SELF, STATIC } from "./Constant.js";
export { Writer } from "./Writer.js";
