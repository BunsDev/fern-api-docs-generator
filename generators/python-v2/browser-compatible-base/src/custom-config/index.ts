export { BasePythonCustomConfigSchema } from "./BasePythonCustomConfigSchema.js";
