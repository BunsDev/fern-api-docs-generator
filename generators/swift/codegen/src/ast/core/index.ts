export { AstNode } from "./AstNode.js";
export { Writer } from "./Writer.js";
