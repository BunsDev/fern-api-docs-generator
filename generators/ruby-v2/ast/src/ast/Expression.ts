import { assertNever } from "@fern-api/core-utils";

import { AstNode } from "./core/AstNode";
import { Writer } from "./core/Writer";

export interface String_ {
    type: "string";
    value: string;
}

export interface Numeric {
    type: "numeric";
    value: number;
}

export interface Boolean_ {
    type: "boolean";
    value: boolean;
}

type InternalExpression = String_ | Numeric | Boolean_;

export class Expression extends AstNode {
    private constructor(public readonly internalExpression: InternalExpression) {
        super();
    }

    public write(writer: Writer): void {
        switch (this.internalExpression.type) {
            case "string":
                writer.write(`"${this.internalExpression.value}"`);
                break;
            case "numeric":
                writer.write(`${this.internalExpression.value}`);
                break;
            case "boolean":
                writer.write(`${this.internalExpression.value}`);
                break;
            default:
                assertNever(this.internalExpression);
        }
    }

    public static string(value: String_["value"]): Expression {
        return new this({ type: "string", value });
    }

    public static numeric(value: Numeric["value"]): Expression {
        return new this({ type: "numeric", value });
    }

    public static boolean(value: Boolean_["value"]): Expression {
        return new this({ type: "boolean", value });
    }
}
