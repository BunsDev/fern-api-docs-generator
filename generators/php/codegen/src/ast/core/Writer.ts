import { AbstractWriter } from "@fern-api/generator-commons";
import { BasePhpCustomConfigSchema } from "../../custom-config/BasePhpCustomConfigSchema";
import { classReference } from "../../php";
import { ClassReference } from "../ClassReference";
import { GLOBAL_NAMESPACE } from "./Constant";

type Namespace = string;

export declare namespace Writer {
    interface Args {
        /* The namespace that is being written to */
        namespace: string;
        /* The root namespace of the project */
        rootNamespace: string;
        /* Custom generator config */
        customConfig: BasePhpCustomConfigSchema;
        /* Import statements */
        references: Record<Namespace, ClassReference>;
    }
}

export class Writer extends AbstractWriter {
    /* The namespace that is being written to */
    public namespace: string;
    /* The root namespace of the project */
    public rootNamespace: string;
    /* Custom generator config */
    public customConfig: BasePhpCustomConfigSchema;
    /* Import statements */
    public readonly references: Record<Namespace, ClassReference> = {};
    public readonly namesWithConflicts = new Set<string>();

    constructor({ namespace, rootNamespace, customConfig, references }: Writer.Args) {
        super();
        this.namespace = namespace;
        this.rootNamespace = rootNamespace;
        this.customConfig = customConfig;
        this.references = references;
        this.namesWithConflicts = this.setConflictingNames(references);
    }

    private setConflictingNames(references: Record<Namespace, ClassReference>): Set<string> {
        const seenNames = new Set<string>();
        return new Set(
            Object.values(references)
                .filter((reference) => {
                    if (seenNames.has(reference.name)) {
                        return true;
                    }
                    seenNames.add(reference.name);
                    return false;
                })
                .map((classReference) => classReference.name)
        );
    }

    public addReference(reference: ClassReference): void {}

    public toString(): string {
        const namespace = `namespace ${this.namespace};`;
        const imports = this.stringifyImports();
        if (imports.length > 0) {
            return `${namespace}

${imports}

${this.buffer}`;
        }
        return namespace + "\n\n" + this.buffer;
    }

    public requiresInlineFullQualification(classReference: ClassReference) {
        return this.namesWithConflicts.has(classReference.name);
    }

    private stringifyImports(): string {
        const classReferences = Object.values(this.references).filter(
            (classReference) =>
                classReference.namespace !== this.namespace && !this.requiresInlineFullQualification(classReference)
        );
        if (classReferences.length === 0) {
            return "";
        }
        let result = classReferences.map((classReference) => `use ${this.toImportString(classReference)};`).join("\n");
        if (result.length > 0) {
            result += "\n";
        }
        return result;
    }

    public toImportString(reference: ClassReference): string {
        return reference.namespace === GLOBAL_NAMESPACE ? reference.name : `${reference.namespace}\\${reference.name}`;
    }
}
