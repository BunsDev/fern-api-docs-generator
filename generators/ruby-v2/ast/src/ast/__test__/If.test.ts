import { ruby } from "../..";
import { BaseRubyCustomConfigSchema } from "../../custom-config/BaseRubyCustomConfigSchema";
import { Expression } from "../Expression";
import { Writer } from "../core/Writer";

describe("If", () => {
    let writerConfig: Writer.Args;

    beforeEach(() => {
        writerConfig = { customConfig: BaseRubyCustomConfigSchema.parse({ clientClassName: "Example" }) };
    });

    describe("write", () => {
        test("writes trailing if", () => {
            const if_ = ruby.if_({ conditional: Expression.boolean(true) });

            expect(if_.toString(writerConfig)).toMatchSnapshot();
        });

        test("writes if", () => {
            const if_ = ruby.if_({
                conditional: Expression.boolean(true),
                consequent: [Expression.string("It works!")]
            });

            expect(if_.toString(writerConfig)).toMatchSnapshot();
        });

        test("writes if/else", () => {
            const if_ = ruby
                .if_({
                    conditional: Expression.boolean(true),
                    consequent: [Expression.string("It works!")]
                })
                .else_({ consequent: [Expression.string("It doesn't work")] });

            expect(if_.toString(writerConfig)).toMatchSnapshot();
        });

        test("writes if/elsif/else", () => {
            const if_ = ruby
                .if_({ conditional: Expression.boolean(true), consequent: [Expression.string("in the if")] })
                .elsif_({ conditional: Expression.numeric(1), consequent: [Expression.string("in the elsif")] })
                .else_({ consequent: [Expression.string("in the else")] });

            expect(if_.toString(writerConfig)).toMatchSnapshot();
        });
    });
});
