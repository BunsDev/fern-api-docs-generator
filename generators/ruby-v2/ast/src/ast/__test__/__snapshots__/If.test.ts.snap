// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`If > write > writes if 1`] = `
"if true
    "It works!"
end"
`;

exports[`If > write > writes if/else 1`] = `
"if true
    "It works!"
else
    "It doesn't work"
end"
`;

exports[`If > write > writes if/elsif/else 1`] = `
"if true
    "in the if"
elsif 1
    "in the elsif"
else
    "in the else"
end"
`;

exports[`If > write > writes trailing if 1`] = `"if true"`;
